### PR TITLE
Add ext specs for map and filter

### PR DESCRIPTION
**Pull request by xldenis**
This adds the missing `extern_spec!` declarations for `map` (we only had `map_inv`) and `filter` (which is new).

### CODE REVIEW

**Comment by jhjourdan** [comment]
There are many remaining warnings visible in CI.

File: creusot-contracts/src/invariant.rs [hunk 1]
```diff
@@ -41,3 +41,8 @@ impl<T: ?Sized> Invariant for &mut T {
 pub fn inv<T: ?Sized>(_: T) -> bool {
     true
 }
+
+#[cfg(not(creusot))]
+pub fn inv<T: ?Sized>(_: &T) -> bool {
+    true
+}
```
Comment (jhjourdan) [inline creusot-contracts/src/invariant.rs:48]:
```suggestion
pub fn inv<T: ?Sized>(_: &T) -> bool {
    panic!()
}
```

File: creusot/tests/should_succeed/iterators/17_filter/why3session.xml [hunk 2]
```diff
@@ -33,6 +34,25 @@
...
+    <proof prover="1"><result status="unknown" time="0.076436" steps="16772"/></proof>
+    <proof prover="2"><result status="timeout" time="1.000000" steps="21843"/></proof>
+    <proof prover="3"><result status="unknown" time="0.033070" steps="10638"/></proof>
+    <transf name="exists" arg1="(empty : seq t_Item&#39;0)">
+     <goal name="vc_produces_refl.0.0.1.0">
+     <proof prover="0"><result status="timeout" time="1.000000" steps="2110361"/></proof>
+     <proof prover="1"><result status="unknown" time="0.074295" steps="16825"/></proof>
+     <proof prover="2"><result status="timeout" time="1.000000" steps="21208"/></proof>
+     <proof prover="3"><result status="unknown" time="0.035550" steps="10695"/></proof>
+     </goal>
```
Comment (jhjourdan) [inline creusot/tests/should_succeed/iterators/17_filter/why3session.xml:52]: You should clean up the proofs!

File: creusot-contracts/src/std/iter/filter.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    #[predicate(prophetic)]
+    fn produces(self, visited: Seq<Self::Item>, succ: Self) -> bool {
+        pearlite! {
+            self.func().unnest(succ.func()) &&
+            // f here is a mapping from indices of `visited` to those of `s`, where `s` is the whole sequence produced by the underlying iterator
+            // Interestingly, Z3 guesses `f` quite readily but gives up *totally* on `s`. However, the addition of the final assertions on the correctness of the values
+            // blocks z3's guess for `f`.
+            exists<s : Seq<Self::Item>, f : Mapping<Int, Int>> self.iter().produces(s, succ.iter()) &&
+                // `f` is a monotone mapping
+                (forall<i : _, j :_ > 0 <= i && i <= j && j < visited.len() ==> 0 <= f.get(i) && f.get(i) <= f.get(j) && f.get(j) < s.len()) &&
```
Comment (jhjourdan) [inline creusot-contracts/src/std/iter/filter.rs:104]:
```suggestion
                (forall<i : _, j :_ > 0 <= i && i < j && j < visited.len() ==> 0 <= f.get(i) && f.get(i) < f.get(j) && f.get(j) < s.len()) &&
```

THIS is injectivity.

File: creusot-contracts/src/std/iter/filter.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                // `f` is injective
+                (forall<i : _, > 0 <= i && i < visited.len() ==> visited[i] == s[f.get(i)]) &&
```
Comment (jhjourdan) [inline creusot-contracts/src/std/iter/filter.rs:106]: That is just saying `visited` is a subsequence of s. This is not injectivity.

File: creusot-contracts/src/std/iter/filter.rs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+where
+    I: Iterator,
+    F: FnMut(&I::Item) -> bool,
+{
+    #[open]
+    #[predicate(prophetic)]
+    fn completed(&mut self) -> bool {
+        pearlite! {
+        (exists<inner : &mut _> inv(inner) && *inner == self.iter() && ^inner == (^self).iter()
+            && inner.completed()) && (*self).func() == (^self).func() }
```
Comment (jhjourdan) [inline creusot-contracts/src/std/iter/filter.rs:91]: I don't see why this is correct. When `Filter::next` returns, the underlying iterator may have performed several steps.

**Comment by xldenis** [comment]
> There are many remaining warnings visible in CI.

I know, there are still some changes i need to make locally.

File: creusot-contracts/src/std/iter/filter.rs (hunk 4 above)
Comment (xldenis) [reply creusot-contracts/src/std/iter/filter.rs:106]: hmm, fair but I wouldn't say its a _subsequence_ here.

File: creusot-contracts/src/std/iter/filter.rs (hunk 5 above)
Comment (xldenis) [reply creusot-contracts/src/std/iter/filter.rs:91]: That's not forbidden?

File: creusot-contracts/src/std/iter/filter.rs (hunk 4 above)
Comment (jhjourdan) [reply creusot-contracts/src/std/iter/filter.rs:106]: That's the definition of the word "subsequence" in mathematics: https://en.wikipedia.org/wiki/Subsequence

The thing you think is a subsequence is a "factor".

File: creusot-contracts/src/std/iter/filter.rs (hunk 5 above)
Comment (xldenis) [reply creusot-contracts/src/std/iter/filter.rs:91]: lol nope, i can prove false trivially, yet no proof fails! unsoundness time!

File: creusot-contracts/src/std/iter/filter.rs (hunk 4 above)
Comment (xldenis) [reply creusot-contracts/src/std/iter/filter.rs:106]: yea but a subsequence in terms of smt sequences is definitely a contiguous series of elements, that's what our `Seq::subsequence` predicate describes.

File: creusot-contracts/src/std/iter/filter.rs (hunk 5 above)
Comment (jhjourdan) [reply creusot-contracts/src/std/iter/filter.rs:91]: Why would it be forbidden? If `Filter` does not find an item that satisfy the predicate, then it returns `None`, even if several items have been produced by the underlying iterator.

Comment (xldenis) [reply creusot-contracts/src/std/iter/filter.rs:91]: agreed, i meant that it wasn't supposed to be forbidden, but now i notice that I never exercised that spec due to an unsoundness in our encoding.

**Pull request merged**